src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,190 @@
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        template<class IteratorT> //TODO is there a way to enforce IteratorT is actually a subclass of std::iterator?
+        SlidingWindow(UInt maxCapacity, IteratorT initialData_begin, IteratorT initialData_end): 
+          SlidingWindow(maxCapacity) {
+  //TODO if there is huge data and small window, we wastefully loop through all data
+  for(IteratorT it = initialData_begin; it != initialData_end; ++it) {
+    append(*it);
+  }
+  idxNext_ = buffer_.size() % maxCapacity;
+        }
+        const UInt maxCapacity;
+        size_t size() const;
+        /** append new value to the end of the buffer and handle the 
+           "overflows"-may pop the first element if full. 
+        */
+        void append(T newValue);
+
+        /** like append, but return the dropped value. isValid indicates 
+            if the return value is valid (not while size()< maxCapacity)
+          :param T newValue - new value to append to the sliding window
+          :param bool isValid - a return pass-by-value that indicates validity
+             of the return T value. for first maxCapacity items it is false, 
+             later always true.
+          :return T dropped value (past oldest element) if isValid; 
+             if not valid, this field holds the oldest value 
+             (but still contained in the window!)
+        */
+        T append(T newValue, bool& isValid);
+
+        /**
+        * :return unordered content (data ) of this sl. window; 
+             call getLinearizedData() if you need them oredered from 
+             oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than 
+            the pure getData() but ensures that the data are ordered (oldest at
+             the beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> getLinearizedData() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  idxNext_ = 0;
+}
+
+
+template<class T>
+size_t SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return buffer_.size();
+}
+
+
+template<class T>
+void SlidingWindow<T>::append(T newValue) {
+  if(size() < maxCapacity) {
+    buffer_.emplace_back(newValue);
```

Comment:
I don't think this is providing any benefit using `emplace_back` rather than `push_back`. `emplace_back` is good for passing constructor parameters directly to the vector, avoiding a copy/move, but here we already have a T instance, and the vector just calls the copy constructor. To really take advantage of this, I think you'd have to add an `emplace` method to the `SlidingWindow` which takes constructor params, and do some crazy templating to make it work.

I might be overlooking something.

If this isn't providing any benefit, we should switch to `push_back` so that it's clear it's not doing any `emplace` magic.